src/nemos/solvers.py

```diff
@@ -0,0 +1,725 @@
+from .typing import KeyArrayLike, Pytree
+
+
+class SVRGState(NamedTuple):
+    iter_num: int
+    key: KeyArrayLike
+    error: float
+    stepsize: float
+    xs: Optional[Pytree] = None
+    df_xs: Optional[Pytree] = None
+
+
+class ProxSVRG:
+    """
+    Prox-SVRG solver
+
+    Borrowing from jaxopt.ProximalGradient, this solver minimizes:
+
+      objective(params, hyperparams_prox, *args, **kwargs) =
+        fun(params, *args, **kwargs) + non_smooth(params, hyperparams_prox)
+
+    Attributes
+    ----------
+    fun: Callable
+        Smooth function of the form ``fun(x, *args, **kwargs)``.
+    prox: Callable
+        Proximal operator associated with the function ``non_smooth``.
+        It should be of the form ``prox(params, hyperparams_prox, scale=1.0)``.
+        See ``jaxopt.prox`` for examples.
+    maxiter : int
+        Maximum number of epochs to run the optimization for.
+    key : jax.random.PRNGkey
+        jax PRNGKey to start with. Used for sampling random data points.
+    stepsize : float
+        Constant step size to use.
+    tol: float
+        Tolerance level for the error when comparing parameters
+        at the end of consecutive epochs to check for convergence.
+    batch_size: int
+        Number of data points to sample per inner loop iteration.
+
+    Example
+    -------
+    def loss_fn(params, X, y):
+        ...
+
+    svrg = ProxSVRG(loss_fn, prox_fun)
+    params, state = svrg.run(init_params, prox_lambda, X, y)
+
+    References
+    ----------
+    1. [Xiao, Lin, and Tong Zhang. "A proximal stochastic gradient method with progressive variance reduction." SIAM Journal on Optimization 24.4 (2014): 2057-2075.](https://arxiv.org/abs/1403.4699v1)
+    2. [Johnson, Rie, and Tong Zhang. "Accelerating stochastic gradient descent using predictive variance reduction." Advances in neural information processing systems 26 (2013).](https://proceedings.neurips.cc/paper/2013/hash/ac1dd209cbcc5e5d1c6e28598e8cbbe8-Abstract.html)
+    """
+
+    def __init__(
+        self,
+        fun: Callable,
+        prox: Callable,
+        maxiter: int = 10_000,
+        key: Optional[KeyArrayLike] = None,
+        stepsize: float = 1e-3,
+        tol: float = 1e-3,
+        batch_size: int = 1,
+    ):
+        self.fun = fun
+        self.maxiter = maxiter
+        self.key = key
+        self.stepsize = stepsize
+        self.tol = tol
+        self.loss_gradient = jit(grad(self.fun))
+        self.batch_size = batch_size
+        self.proximal_operator = prox
+
+    def init_state(
+        self,
+        init_params: Pytree,
+        hyperparams_prox: Any,
+        *args,
+        init_full_gradient: bool = False,
+    ) -> SVRGState:
+        """
+        Initialize the solver state
+
+        Parameters
+        ----------
+        init_params :
+            Pytree containing the initial parameters.
+            For GLMs it's a tuple of (W, b)
+        hyperparams_prox :
```

Comment:
since this is unused, should we raise an error or warning if it gets set?

Reply:
What do you mean by it getting set?

[This is how jaxopt treats this](https://github.com/google/jaxopt/blob/6d882923a54feb6f68b46193d9787cc73bfdadd3/jaxopt/_src/proximal_gradient.py#L179). I wanted to be consistent with them.

Reply:
I think that's fine, but what I meant was that if these are specified, we should raise a warning or error. as in `if hyperparams_prox is not None`. but actually, I think I'm confused -- why isn't this argument optional?

Reply:
unsure why, but in the general `jaxopt.ProximalGradient` API, both `init_state` and `update` require a `hyperparameter_prox` input. 

This is an annoying choice, they should have make it keyword and optional, but in their case it made sense to have an hyperparameter variable at state initialization since you cannot anticipate if a specific proximal method would require it or not.

For us, I don't think we should care about this parameter for this ProxSVRG since we are not implementing a base class for any stochastic solver (which we should, but not in this PR). I'll remove it.

An idea for a refinement of this solver class is to have a general interface for stochastic solvers (with inner/outer loop structure ) and the inner loop should be an abstract class, as well as equipping the class with a `run_iterator` for batching. I'll add this idea to Bence's issue #194.